File: ldk-bukkit-nbt/src/main/java/com/lgou2w/ldk/bukkit/nbt/NBTFactory.java
/*
 * Copyright (C) 2016-2021 The lgou2w <lgou2w@hotmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.bukkit.nbt;

import com.lgou2w.ldk.nbt.BaseTag;
import com.lgou2w.ldk.nbt.ByteTag;
import com.lgou2w.ldk.nbt.ByteArrayTag;
import com.lgou2w.ldk.nbt.CompoundTag;
import com.lgou2w.ldk.nbt.DoubleTag;
import com.lgou2w.ldk.nbt.EndTag;
import com.lgou2w.ldk.nbt.FloatTag;
import com.lgou2w.ldk.nbt.IntTag;
import com.lgou2w.ldk.nbt.IntArrayTag;
import com.lgou2w.ldk.nbt.ListTag;
import com.lgou2w.ldk.nbt.LongTag;
import com.lgou2w.ldk.nbt.LongArrayTag;
import com.lgou2w.ldk.nbt.ShortTag;
import com.lgou2w.ldk.nbt.StringTag;
import com.lgou2w.ldk.nbt.TagType;
import com.lgou2w.ldk.reflect.ConstructorAccessor;
import com.lgou2w.ldk.reflect.ConstructorReflectionMatcher;
import com.lgou2w.ldk.reflect.DataType;
import com.lgou2w.ldk.reflect.FieldAccessor;
import com.lgou2w.ldk.reflect.FuzzyReflection;
import com.lgou2w.ldk.reflect.MethodAccessor;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Supplier;

import static com.lgou2w.ldk.bukkit.reflect.MinecraftReflection.getMinecraftClass;
import static com.lgou2w.ldk.bukkit.reflect.MinecraftReflection.getMinecraftClassOrNull;

public final class NBTFactory {

  private NBTFactory() { }

  /// NMS Classes

  @NotNull public final static Class<?> CLASS_NBT_BASE;
  @NotNull public final static Class<?> CLASS_NBT_TAG_END;
  @NotNull public final static Class<?> CLASS_NBT_TAG_LIST;
  @NotNull public final static Class<?> CLASS_NBT_TAG_COMPOUND;
  @Nullable public final static Class<?> CLASS_NBT_TAG_LONG_ARRAY; // since Minecraft 1.12

  static {
    try {
      CLASS_NBT_BASE = getMinecraftClass("NBTBase");
      CLASS_NBT_TAG_END = getMinecraftClass("NBTTagEnd");
      CLASS_NBT_TAG_LIST = getMinecraftClass("NBTTagList");
      CLASS_NBT_TAG_COMPOUND = getMinecraftClass("NBTTagCompound");
      CLASS_NBT_TAG_LONG_ARRAY = getMinecraftClassOrNull("NBTTagLongArray");
    } catch (ClassNotFoundException e) {
      throw new RuntimeException("Error in initializing NBTFactory internal static block:", e);
    }
  }

  /// NMS Accessors

  // NMS.NBTTagList -> private byte type;
  final static Supplier<@NotNull FieldAccessor<Object, Byte>> FIELD_NBT_TAG_LIST_TYPE
    = FuzzyReflection.lazySupplier(CLASS_NBT_TAG_LIST, true, fuzzy -> fuzzy
    .useFieldMatcher()
    .withoutModifiers(Modifier.FINAL, Modifier.STATIC)
    .withType(byte.class)
    .resultAccessorAs("Missing match: NMS.NBTTagList -> Field: byte type"));

  // NMS.NBTBase -> public abstract byte getTypeId();
  final static Supplier<@NotNull MethodAccessor<Object, Byte>> METHOD_NBT_BASE_GET_TYPE_ID
    = FuzzyReflection.lazySupplier(CLASS_NBT_BASE, true, fuzzy -> fuzzy
    .useMethodMatcher()
    .withModifiers(Modifier.PUBLIC, Modifier.ABSTRACT)
    .withArgsCount(0)
    .withType(byte.class)
    .resultAccessorAs("Missing match: NMS.NBTBase -> Method: public abstract byte getTypeId()"));

  final static Map<TagType, FieldAccessor<Object, Object>> INTERNAL_FIELD_MAP = new HashMap<>();
  final static Map<TagType, ConstructorAccessor<Object>> INTERNAL_CONSTRUCTOR_MAP = new HashMap<>();
  final static Map<TagType, String> INTERNAL_CLASS_NAME_MAP = Collections.unmodifiableMap(new HashMap<TagType, String>() {{
    put(TagType.END,        "NBTTagEnd");
    put(TagType.BYTE,       "NBTTagByte");
    put(TagType.SHORT,      "NBTTagShort");
    put(TagType.INT,        "NBTTagInt");
    put(TagType.LONG,       "NBTTagLong");
    put(TagType.FLOAT,      "NBTTagFloat");
    put(TagType.DOUBLE,     "NBTTagDouble");
    put(TagType.STRING,     "NBTTagString");
    put(TagType.BYTE_ARRAY, "NBTTagByteArray");
    put(TagType.INT_ARRAY,  "NBTTagIntArray");
    put(TagType.LIST,       "NBTTagList");
    put(TagType.COMPOUND,   "NBTTagCompound");
    put(TagType.LONG_ARRAY, "NBTTagLongArray");
  }});

  static ConstructorAccessor<Object> lookupInternalConstructor(TagType type) throws ClassNotFoundException {
    if (type == TagType.END) throw new IllegalArgumentException("END");
    ConstructorAccessor<Object> accessor = INTERNAL_CONSTRUCTOR_MAP.get(type);
    if (accessor == null) {
      @NotNull String className = INTERNAL_CLASS_NAME_MAP.get(type);
      ConstructorReflectionMatcher<Object> matcher = FuzzyReflection
        .of(getMinecraftClass(className), true)
        .useConstructorMatcher();
      accessor = (
        type.isListOrCompound()
          ? matcher.withArgsCount(0)
          : matcher.withArgs(type.getPrimitive())
      ).resultAccessor("Missing match: NMS." + className + " -> Constructor(" + (!type.isListOrCompound()
        ? type.getPrimitive().getSimpleName() + ")"
        : ")"
      ));
      INTERNAL_CONSTRUCTOR_MAP.put(type, accessor);
    }
    return accessor;
  }

  static FieldAccessor<Object, Object> lookupInternalField(TagType type) throws ClassNotFoundException {
    if (type == TagType.END) throw new IllegalArgumentException("END");
    FieldAccessor<Object, Object> accessor = INTERNAL_FIELD_MAP.get(type);
    if (accessor == null) {
      @NotNull String className = INTERNAL_CLASS_NAME_MAP.get(type);
      accessor = FuzzyReflection.of(getMinecraftClass(className), true)
        .useFieldMatcher()
        .withType(type.getPrimitive())
        .resultAccessor("Missing match: NMS." + className + " -> Field: " + type.getPrimitive().getSimpleName() + " value");
      INTERNAL_FIELD_MAP.put(type, accessor);
    }
    return accessor;
  }

  // NMS.NBTTagEnd -> public final static NMS.NBTTagEnd instance; or new instance
  final static Object INSTANCE_NBT_TAG_END = lookupEndInstanceOrNew();

  static Object lookupEndInstanceOrNew() {
    FuzzyReflection reflection = FuzzyReflection.of(CLASS_NBT_TAG_END, true);
    List<FieldAccessor<Object, Object>> fields = reflection
      .useFieldMatcher()
      .withModifiers(Modifier.PUBLIC, Modifier.FINAL, Modifier.STATIC)
      .withType(CLASS_NBT_TAG_END)
      .resultAccessors();
    if (!fields.isEmpty()) {
      return fields.get(0).get(null);
    } else {
      return reflection
        .useConstructorMatcher()
        .resultAccessor("Missing match: NMS.NBTTagEnd -> Constructor()")
        .newInstance();
    }
  }

  /// Public API

  @NotNull
  @Contract("null -> fail; !null -> !null")
  @SuppressWarnings({ "unchecked", "ConstantConditions" })
  public static BaseTag<?> from(Object nbt) throws IllegalArgumentException {
    Objects.requireNonNull(nbt, "nbt");
    if (!CLASS_NBT_BASE.isInstance(nbt))
      throw new IllegalArgumentException("Value type of the instance does not match. (Expected: " + CLASS_NBT_BASE + ")");
    int typeId = METHOD_NBT_BASE_GET_TYPE_ID.get().invoke(nbt);
    TagType type = Objects.requireNonNull(TagType.fromId(typeId), "Invalid nbt type id: " + typeId);
    FieldAccessor<Object, Object> field;
    try {
      field = lookupInternalField(type);
    } catch (ClassNotFoundException e) {
      throw new UnsupportedOperationException("Server version does not support this type: " + type);
    }
    Object value = Objects.requireNonNull(field.get(nbt), "Null value for nbt: " + nbt);
    switch (type) {
      case END: return EndTag.INSTANCE;
      case BYTE: return new ByteTag((byte) value);
      case SHORT: return new ShortTag((short) value);
      case INT: return new IntTag((int) value);
      case LONG: return new LongTag((long) value);
      case FLOAT: return new FloatTag((float) value);
      case DOUBLE: return new DoubleTag((double) value);
      case BYTE_ARRAY: return new ByteArrayTag((byte[]) value);
      case STRING: return new StringTag((String) value);
      case INT_ARRAY: return new IntArrayTag((int[]) value);
      case LONG_ARRAY: return new LongArrayTag((long[]) value);
      case LIST:
        ListTag list = new ListTag();
        List<Object> listValue = (List<Object>) value;
        for (Object element : listValue) {
          BaseTag<?> elementFrom = from(element);
          list.add(elementFrom);
        }
        return list;
      case COMPOUND:
        CompoundTag compound = new CompoundTag();
        Map<String, Object> compoundValue = (Map<String, Object>) value;
        for (Map.Entry<String, Object> entry : compoundValue.entrySet()) {
          BaseTag<?> entryValueFrom = from(entry.getValue());
          compound.set(entry.getKey(), entryValueFrom);
        }
        return compound;
      default:
        throw new UnsupportedOperationException();
    }
  }

  @NotNull
  @Contract("null -> fail; !null -> !null")
  public static Object to(BaseTag<?> nbt) throws IllegalArgumentException {
    Objects.requireNonNull(nbt, "nbt");
    switch (nbt.getType()) {
      case END: return INSTANCE_NBT_TAG_END;
      case LIST:
        ListTag list = (ListTag) nbt;
        List<Object> listValue = new ArrayList<>();
        for (BaseTag<?> element : list) {
          Object elementTo = to(element);
          listValue.add(elementTo);
        }
        return newInternal(TagType.LIST, listValue, list.getElementType());
      case COMPOUND:
        CompoundTag compound = (CompoundTag) nbt;
        Map<String, Object> compoundValue = new LinkedHashMap<>();
        for (Map.Entry<String, BaseTag<?>> entry : compound.entrySet()) {
          Object entryValueTo = to(entry.getValue());
          compoundValue.put(entry.getKey(), entryValueTo);
        }
        return newInternal(TagType.COMPOUND, compoundValue);
      case LONG_ARRAY:
        boolean supported = CLASS_NBT_TAG_LONG_ARRAY != null;
        if (supported) return newInternal(nbt.getType(), nbt.getValue());
        // Server does not support, convert to list to compatibility
        LongArrayTag longArray = (LongArrayTag) nbt;
        ListTag longList = new ListTag();
        longList.addLong(longArray.getValue());
        return to(longList);
      default:
        return newInternal(nbt.getType(), nbt.getValue());
    }
  }

  @NotNull
  @Contract("null, _ -> fail")
  public static Object newInternal(
    TagType type,
    @Nullable Object value
  ) throws IllegalArgumentException {
    return newInternal(type, value, TagType.END);
  }

  @NotNull
  @Contract("null, _, _ -> fail; _, _, null -> fail")
  public static Object newInternal(
    TagType type,
    @Nullable Object value,
    TagType listElementType
  ) throws IllegalArgumentException {
    Objects.requireNonNull(type, "type");
    Objects.requireNonNull(listElementType, "listElementType");
    if (value != null && !type.getPrimitive().isAssignableFrom(DataType.ofPrimitive(value.getClass())))
      throw new IllegalArgumentException("Value '" + value.getClass() + "' and type mismatch. (Expected: " + type.getPrimitive() + ")");
    ConstructorAccessor<Object> constructor;
    FieldAccessor<Object, Object> field;
    try {
      constructor = lookupInternalConstructor(type);
      field = lookupInternalField(type);
    } catch (ClassNotFoundException e) {
      throw new UnsupportedOperationException("Server version does not support this type: " + type);
    }
    if (type.isListOrCompound()) {
      Object inst = constructor.newInstance();
      if (value != null) field.set(inst, value);
      if (value != null && type == TagType.LIST)
        FIELD_NBT_TAG_LIST_TYPE.get().set(inst, (byte) listElementType.getId());
      return inst;
    } else {
      return constructor.newInstance(value);
    }
  }
}

File: ldk-bukkit/ldk-bukkit-common/src/main/kotlin/com/lgou2w/ldk/bukkit/nbt/NBTFactory.kt
/*
 * Copyright (C) 2018 The lgou2w (lgou2w@hotmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.bukkit.nbt

import com.lgou2w.ldk.bukkit.reflect.MinecraftReflection
import com.lgou2w.ldk.bukkit.reflect.lazyMinecraftClass
import com.lgou2w.ldk.nbt.*
import com.lgou2w.ldk.reflect.AccessorField
import com.lgou2w.ldk.reflect.AccessorMethod
import com.lgou2w.ldk.reflect.FuzzyReflect
import com.lgou2w.ldk.reflect.Visibility
import java.io.DataInput
import java.io.DataOutput
import java.util.*

object NBTFactory {

    @JvmStatic val CLASS_NBT_BASE by lazyMinecraftClass("NBTBase")
    @JvmStatic val CLASS_NBT_TAG_LIST by lazyMinecraftClass("NBTTagList")
    @JvmStatic val CLASS_NBT_TAG_COMPOUND by lazyMinecraftClass("NBTTagCompound")
    @JvmStatic val CLASS_NBT_STREAM by lazyMinecraftClass("NBTCompressedStreamTools")
    @JvmStatic val CLASS_NBT_READ_LIMITER by lazyMinecraftClass("NBTReadLimiter")

    // NMS.NBTList -> private byte type
    @JvmStatic val NBT_LIST_TYPE_FIELD: AccessorField<Any, Byte> by lazy {
        FuzzyReflect.of(CLASS_NBT_TAG_LIST, true)
            .useFieldMatcher()
            .withType(Byte::class.java)
            .resultAccessorAs<Any, Byte>()
    }

    // NMS.NBTBase -> public static NBTBase createTag(byte)
    @JvmStatic val METHOD_NBT_CREATE: AccessorMethod<Any, Any> by lazy {
        FuzzyReflect.of(CLASS_NBT_BASE, true)
                .useMethodMatcher()
                .withVisibilities(Visibility.STATIC)
                .withName("createTag")
                .withType(CLASS_NBT_BASE)
                .withParams(Byte::class.java)
                .resultAccessor()
    }

    // NMS.NBTBase -> public abstract byte getTypeId()
    @JvmStatic val METHOD_NBT_GET_TYPE_ID: AccessorMethod<Any, Byte> by lazy {
        FuzzyReflect.of(CLASS_NBT_BASE, true)
                .useMethodMatcher()
                .withVisibilities(Visibility.PUBLIC, Visibility.ABSTRACT)
                .withName("getTypeId")
                .withType(Byte::class.java)
                .resultAccessorAs<Any, Byte>()
    }

    // NMS.NBTCompressedStreamTools -> public static NMS.NBTBase read(DataInput, Int, NMS.NBTReadLimiter)
    @JvmStatic val MEHTOD_NBT_READ: AccessorMethod<Any, Any> by lazy {
        FuzzyReflect.of(CLASS_NBT_STREAM, true)
                .useMethodMatcher()
                .withVisibilities(Visibility.STATIC)
                .withType(CLASS_NBT_BASE)
                .withParams(DataInput::class.java, Int::class.java, CLASS_NBT_READ_LIMITER)
                .resultAccessor()
    }


    // NMS.NBTCompressedStreamTools -> public static void write(NMS.NBTBase, DataOutput)
    @JvmStatic val METHOD_NBT_WRITE: AccessorMethod<Any, Any> by lazy {
        FuzzyReflect.of(CLASS_NBT_STREAM, true)
                .useMethodMatcher()
                .withVisibilities(Visibility.STATIC)
                .withType(Void::class.java)
                .withParams(CLASS_NBT_BASE, DataOutput::class.java)
                .resultAccessor()
    }

    @JvmStatic
    fun fromNMS(nms: Any?): NBTBase<*>? {
        if (nms == null)
            return null
        val type = NBTType.fromId(METHOD_NBT_GET_TYPE_ID.invoke(nms)!!.toInt())!!
        @Suppress("UNCHECKED_CAST")
        return when (type) {
            NBTType.TAG_COMPOUND -> {
                val compound = NBTTagCompound()
                val value = NBT_TYPE_FIELD(type)[nms] as Map<String, Any>
                value.entries.forEach { compound[it.key] = fromNMS(it.value)!! }
                compound
            }
            NBTType.TAG_LIST -> {
                val list = NBTTagList()
                val value = NBT_TYPE_FIELD(type)[nms] as List<Any>
                value.forEach { list.add(fromNMS(it)!!) }
                list
            }
            NBTType.TAG_END -> NBTTagEnd.INSTANCE
            NBTType.TAG_BYTE -> NBTTagByte(NBT_TYPE_FIELD(type)[nms] as Byte)
            NBTType.TAG_SHORT -> NBTTagShort(NBT_TYPE_FIELD(type)[nms] as Short)
            NBTType.TAG_INT -> NBTTagInt(NBT_TYPE_FIELD(type)[nms] as Int)
            NBTType.TAG_LONG -> NBTTagLong(NBT_TYPE_FIELD(type)[nms] as Long)
            NBTType.TAG_FLOAT -> NBTTagFloat(NBT_TYPE_FIELD(type)[nms] as Float)
            NBTType.TAG_DOUBLE -> NBTTagDouble(NBT_TYPE_FIELD(type)[nms] as Double)
            NBTType.TAG_BYTE_ARRAY -> NBTTagByteArray(NBT_TYPE_FIELD(type)[nms] as ByteArray)
            NBTType.TAG_STRING -> NBTTagString(NBT_TYPE_FIELD(type)[nms] as String)
            NBTType.TAG_INT_ARRAY -> NBTTagIntArray(NBT_TYPE_FIELD(type)[nms] as IntArray)
            else -> throw UnsupportedOperationException()
        }
    }

    @JvmStatic
    fun toNMS(nbt: NBTBase<*>?): Any? {
        if (nbt == null)
            return null
        return when (nbt.type) {
            NBTType.TAG_END -> NBT_END_INSTANCE
            NBTType.TAG_COMPOUND -> {
                val value = (nbt as NBTTagCompound).entries.associate { it.key to toNMS(it.value) }
                createInternal(NBTType.TAG_COMPOUND, value)
            }
            NBTType.TAG_LIST -> {
                val value = (nbt as NBTTagList).map { toNMS(it) }
                val handle = createInternal(NBTType.TAG_LIST, value)
                NBT_LIST_TYPE_FIELD[handle] = nbt.elementType.id.toByte()
                handle
            }
            else -> createInternal(nbt.type, nbt.value)
        }
    }

    @JvmStatic
    fun createInternal(type: NBTType, value: Any? = null) : Any {
        val instance = METHOD_NBT_CREATE.invoke(null, type.id.toByte())
        val valueAccessor = NBT_TYPE_FIELD(type)
        if (value != null)
            valueAccessor[instance] = value
        return instance as Any
    }

    private val NBT_END_INSTANCE: Any by lazy {
        FuzzyReflect.of(MinecraftReflection.getMinecraftClass("NBTTagEnd"), true)
                .useConstructorMatcher()
                .resultAccessor()
                .newInstance()
    }

    private val NBT_TYPE_FIELD_ACCESSORS: MutableMap<NBTType, AccessorField<Any, Any>> = HashMap()
    private val NBT_TYPE_FIELD: (NBTType) -> AccessorField<Any, Any> = { type ->
        var accessor = NBT_TYPE_FIELD_ACCESSORS[type]
        if (accessor == null) {
            if (type == NBTType.TAG_END)
                throw IllegalArgumentException("类型 TAG_END 没有值成员字段.")
            accessor = FuzzyReflect.of(MinecraftReflection.getMinecraftClass(when (type) {
                NBTType.TAG_END -> "NBTTagEnd"
                NBTType.TAG_BYTE -> "NBTTagByte"
                NBTType.TAG_SHORT -> "NBTTagShort"
                NBTType.TAG_INT -> "NBTTagInt"
                NBTType.TAG_LONG -> "NBTTagLong"
                NBTType.TAG_FLOAT -> "NBTTagFloat"
                NBTType.TAG_DOUBLE -> "NBTTagDouble"
                NBTType.TAG_STRING -> "NBTTagString"
                NBTType.TAG_BYTE_ARRAY -> "NBTTagByteArray"
                NBTType.TAG_INT_ARRAY -> "NBTTagIntArray"
                NBTType.TAG_LIST -> "NBTTagList"
                NBTType.TAG_COMPOUND -> "NBTTagCompound"
            }), true)
                    .useFieldMatcher()
                    .withType(type.primitive)
                    .resultAccessor()
            NBT_TYPE_FIELD_ACCESSORS[type] = accessor
        }
        accessor
    }
}
